### ISSUE TITLE
Pull queue, agent implementation

### ISSUE THREAD

**Issue by bradrydzewski** [drone/drone#1386]
### update

This issue will be closed once the communication protocol is finalized. We are currently looking to implement the STOMP protocol with features based on Apache Apollo (http://activemq.apache.org/apollo/documentation/stomp-manual.html). This will provide a standard communication protocol with the potential to easily choose between the embedded drone queue and a compatible STOMP mq server for added scaleability.

* [x] implement stomp protocol for standard communication
* [x] update documentation to reflect protocol change and configuration
* [x] remove existing custom protocol
* [x] remove existing agent token / secret for custom protocol

---

This is a proposal to switch from a push queue in 0.4 to a pull queue in 0.5. The current version of Drone manages incoming build requests, and pushes builds to docker daemons when build slots are available. This doesn't really allow for very fine-grained scheduling, and using the Docker remote API as the transport protocol makes supporting non-Docker builds (like iOS app) much more difficult.

The pull queue would have us place an agent on a server that will poll the main Drone server for builds. The agent would be started on each server by running the Drone server in agent mode:

```
docker run drone/drone --agent --host=http://mydrone.com
```

The Drone server would need an endpoint for polling jobs:

```
POST /api/work/pull
```

And an endpoint for posting job results, including intermediate updates like pending, running, finished, as well as build logs:

```
POST /api/work
```

Below discusses some pros and cons to the current approach vs. proposed new approach.

### Durability

The Drone server can be taken down for periodic upgrades without worrying about losing the in-memory build queue (see https://github.com/drone/drone/issues/1195). Agents can use backoff to retry contacting Drone server if unavailable when posting results or requesting jobs.

### Multi-architecture

This makes it easier to support multiple operating system and architectures, especially those that don't support Docker (I'm looking at you, iOS). A build system that doesn't support iOS applications can never be a true alternative to Jenkins.

### Resource Usage

This allows (in a future release) the agent to regulate how many builds can execute based on resource usage, instead of just a fixed number (ie 2 builds on this server). This is similar to how Docker swarm has different profiles for allocating resources, and we could (eventually) provide something similar.

### Elasticity

The current design makes increasing or decreasing capacity relatively simple, however, I think agents make this even easier. You could simply snapshot an image or AMI and start up a bunch of servers in batch, which would immediately start processing jobs when they come online. This could still be used in conjunction with Docker machine to bring up the server and run an agent. It would also work better with something like Rancher and Mesos, which have their own GUIs for launching servers.

### Extra Setup

This adds an extra step to our installation. Right now you can simply start Drone and use the host machines Docker daemon to run builds. I want to make sure Drone is easy to evaluate, but I do think agents will scale better for real-world, multi-server installs.

### Secure Communication

We need to find a way to secure communications between the Drone server and their agents. Docker handles this today by creating and installing certificates (using Docker machine) and using TCP for communications.

### Single Server Installs

Running Drone on a single server (with no external build servers) may be slightly more complex because we'll need to deal with inter-container communication.

## Support Both?

While technically possible, I want to pick one approach and execute and document it extremely well. Attempting to support both will likely result in more complex code, complex documentation and confusing and increased support for the community.

**Comment by gtaylor** [comment]
While pull-based agents would (as you noted) make setup more involved, I think the pros probably outweigh the cons.

Some things that appeal to me:
* Not having to open sockets on all of the worker machines. I'd rather focus my security efforts on the Drone server as much as possible.
* Might end up being a little less fragile.
* If I am understanding correctly, container orchestration systems like Kubernetes will pretty much automatically get agent [auto-scaling](http://kubernetes.io/v1.1/docs/user-guide/horizontal-pod-autoscaler.html) support. That's code that we don't have to write and maintain, and it lets admins adapt their Drone configuration to the paradigms and patterns they are most familiar and comfortable with. Did I mention that we don't have to write code for this, too?
* Might be easier to write custom agents that have alternative behaviors this way. Meet the pull-based "interface" and go nuts. At the very least, the development flow might be easier since you've got a more full separation from Drone core.

Things to think about or build in mind with:
* A build gets pulled by an agent. Work starts, but for any number of reasons the agent dies without reporting. We'll want to make sure the build isn't stuck in an "In Progress" state indefinitely, since the agent died. Will be crucial to nail this for general reliability and in case an aggressive auto-scaler kills a container/machine that's doing work without waiting for it to finish.
* Multi-agent build parallelism. I think this is an important feature for modern CI setups to do well with. Given a build, launch the same build on multiple agents with a different evar/build var set. Let the user figure out the rest. CircleCI [does this well](https://circleci.com/docs/parallel-manual-setup#manual-balancing), IMO. Give them the ability to do it, let them figure out how to split the work up. We obviously don't need a solution in place anytime soon, but I'd hope that we have a rough ideal-case design or two sketched out to make sure that our pull-based implementation allows us to eventually tackle this.
* As you mentioned in Gitter, matrix builds. This and build parallelism remind me of Celery's notion of [chords](http://docs.celeryproject.org/en/latest/reference/celery.html#celery.chord). All of the sub-tasks must finish before the top-level job is marked as complete.

I like the idea of the switch, overall. We can probably mitigate some of the setup overhead with documentation and maybe an *option* to run drone + the agent in a single container for evaluation or low-traffic setups.

**Comment by bradrydzewski** [comment]
@gtaylor agree with everything you've written above

> Not having to open sockets on all of the worker machines. I'd rather focus my security efforts on the Drone server as much as possible.

We may need to open a port on the worker machine to stream build logs. When a client visits the page of a running build, we create a websocket to the server, which would connect to the worker machine and proxy the logs.

An alternative design would be for the worker to send the log stream to the server, to avoid opening a port. The server then would multi-cast the logs to clients that may connect. The drawbacks to this approach are that a) the server would need to store the logs in-memory for every running build which could limit server scalability and b) this makes clustering the server difficult since we would need a central cache for the logs.

But if we can work around those two issues, it would be nice to avoid bi-direction communication ...

**Comment by macb** [comment]
It might be worth pointing out that there would need to be some additional logic to prevent 2 workers from picking up the same job in parallel. This seems like it'd tie in with the "a worker dies unexpectedly" story.

What would be a unit of work to be picked up? Would it be a single job? Or a build? How would this tie in with matrix builds?

**Comment by bradrydzewski** [comment]
@macb thanks for joining the conversation. I've attempted answered your questions below. I'd love to know your overall thoughts on the proposal? If we can implement this correctly, and keep the setup simple, do you see this as an improvement?

> It might be worth pointing out that there would need to be some additional logic to prevent 2 workers from picking up the same job in parallel.

We can use a Go channel to queue up jobs. The agents could long poll for jobs and connect to the channel to receive the jobs. This is similar to what we do in 0.4, however, the 0.4 server also manages execution state and lifecycle which is why restoring everyone when the server restarts is a pain. This proposed approach will allow us to restore the queue without having to restore execution state, which is much more simple.

> This seems like it'd tie in with the "a worker dies unexpectedly" story.

I'm thinking we can have agents ping the server every N minutes. If an agent doesn't ping within N minutes + some buffer we can re-allocate the job to another server. It sounds simple enough in my head but this this the first time I've through about it :smile: 

> What would be a unit of work to be picked up? Would it be a single job? Or a build? How would this tie in with matrix builds?

Great question. Matrix builds make this a bit more complicated because we need to execute notifications after the all matrix jobs are complete. I think we'll need a unit of work that is a bit more flexible, and perhaps a priority queue to always ensure notifications are executed after build completion in a timely manner.

I don't have an immediate answer for this, but will give it some thought and post back.

**Comment by gtaylor** [comment]
It sounds like there is some overlap between matrix and parallel builds. Solve one and the other becomes a smaller task.

Re: The queue stuff. There are a ton of queue daemons that can do this stuff, gotta figure someone else has already done the work to expose a queue library with timeouts and retries.

**Comment by nlf** [comment]
in general, i like the idea of moving to a pull queue very much.

my big concerns are things that have already been brought up.. what happens if two workers try to grab the same job at the same time? this can definitely be mitigated, but it's a challenge to make sure you think about.

also why the opposition to bidirectional communication? it seems to me like it's more or less necessary for things like log streaming and notifications. a simple websocket server/client setup would make all of this much easier. when a new build needs to run, emit an event (potentially with constraints that would be resolved on the client side), clients could then send a response accepting the job (which would only work once, later ACKs would be rejected so the client is aware the job is already taken). logs could be streamed over the socket as well, and it still doesn't require opening any ports on the client side of things.

**Comment by bradrydzewski** [comment]
thanks @nlf for the comments. I also wanted to document your suggestion from Gitter for keeping the logs out of RAM. The agent can stream the logs to the server, the server can write the logs to a tmp file, and connected clients can stream the output from the tmp file

**Comment by jackspirou** [comment]
A couple thoughts: 

* I agree with @nlf that bidirectional communication between the server and workers is necessary (cancel a job and such).

* :+1: for the idea of the server to stream logs to a tmp file and then that file output gets streamed to the client.
   * But... couldn't the client also directly open a websocket to the worker?  The server could tell the client which worker to open a connection with.  Workers could also write the output to a db if needed while they stream results to the client.  Just a crazy thought.

* While the proposed design is great for manual or auto scaling of drone, it does make single installations more complex (as mentioned above).  I agree that having two architectures is not ideal, but what about two different installation scripts or methods?  Would that even make sense? 

* I think mitigating two workers picking up the same build could be done in many different ways... some more complex than others. 
  * A silly thought, but raft consensus would be interesting and might allow scaling out of more than one drone server behind a load balancer.  Also it may not practical for this sort of problem (way over kill?).

* Secure server communication is a concern for me. Who provides the certs? How do you add a worker to the pool securely? Repo secrets passed across networks and proxies via SSL/TLS (don't anger fragile/touchy corporate proxies!).  Do we have a mode without SSL/TLS to debug silly internally networks?

Overall I think the new design proposal is an exciting next step for drone, but poses new challenges we all have mentioned thus far.  I think the benefits far outweigh the costs for a VERY scalable drone solution that can tackle iOS and other builds.

**Comment by gtaylor** [comment]
> But... couldn't the client also directly open a websocket to the worker? The server could tell the client which worker to open a connection with

I wouldn't be crazy about this, since now we're talking about exposing ports on all of our worker VMs instead of just the Drone server (if I am understanding correctly).

> While the proposed design is great for manual or auto scaling of drone, it does make single installations more complex (as @bradrydzewski mentioned above).

We can preserve the existing simple install by doing something like the GitLab Omnibus image does, at least by default. Those with more traffic or rigorous conditions could run separate containers for Drone and its workers.

Also, I'll leave this here for worker discovery: https://www.serfdom.io/

**Comment by tboerger** [comment]
I really like most ideas mentioned above. But I would prefer to let the server only communicate to the clients, including a load balancer in front of multiple servers. A socket connection between servers and workers sounds reasonable. For log writing and streaming I would suggest a database because temp files won't work for loadbalanced servers. For discovery of workers I like the idea of using serf :)

**Comment by bradrydzewski** [comment]
I've been trying to think of alternatives to workers automatically streaming the logs to the build server. The reason is that, in most cases, this is un-necessary unless there is a user on the website attempting to view the running build.

I'm worried that if we automatically stream the logs for every build we'll hit network limitations and filesystem limitations for larger installs (1000+ workers).

Instead I'm thinking when a user attempts to view the logs for a running build it opens a websocket to the Drone server that blocks and waits. The agent is notified that a user wants the log stream, so the starts pushing the to the server to the server. The agent log stream is somehow connected to the block user websocket, which immediately starts an `io.Copy` to send the data back to the client.

Not sure if I described it correctly, but I believe it would still allow uni-directional communication from agent to server via websocket, without less risk of overloading the server with too many build streams or having to worry about disk performance and temp files.

**Comment by gtaylor** [comment]
Just to be sure: I'd still be able to look at all of my historical build logs (at least the latest X builds or days worth) after the fact? We find ourselves checking our old build logs periodically for various reasons (timings, output differences, watching some moving parts that we haven't factored out yet).

**Comment by bradrydzewski** [comment]
@gtaylor yes! at the end of the build, the agent will always post the updated build information and full build logs to the drone server to get persisted in the database.

The on-demand logs described above only impacts real-time streaming

**Comment by gtaylor** [comment]
Cool cool cool

**Comment by Spedge** [comment]
Has there been any update on this ticket? We're kinda concerned about the durability aspect you've mentioned in the ticket and are just making the assumption that this work is starting to be investigated or even actioned :)

**Comment by bradrydzewski** [comment]
@Spedge I wouldn't be too overly concerned about the durability aspect unless you have a high volume drone install and plan on stopping / starting your drone instance during working hours. If you do stop / start drone while builds are running, drone will mark those builds as failed on restart, but will leave behind artifacts (build containers) which will need to be manually removed.

So just to clarify, restarting the drone instance does not result in data loss. Restarting drone results in previously pending or running builds to be force-marked as failed and then some manual cleanup is required.

**Comment by xiaods** [comment]
the issue is outdate?

**Comment by tboerger** [comment]
No, it's not outdated. The communication protocol is not finished yet

**Comment by bradrydzewski** [comment]
closing issue, agent code and protocol updates merged to master
